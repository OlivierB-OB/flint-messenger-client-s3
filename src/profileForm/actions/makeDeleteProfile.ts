import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { history } from '../../history';
import { updateProfileFormStatus } from './updateProfileFormStatus';

export function makeDeleteProfile() {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    
    // FIXME confirm before sending

    dispatch(updateProfileFormStatus('unavailable'));

    try {
      await axios.delete(`${process.env.REACT_APP_BACKEND}/profile`, { withCredentials: true });
      history.push(`/login`);
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
}
