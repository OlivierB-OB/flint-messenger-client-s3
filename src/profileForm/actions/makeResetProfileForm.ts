import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { resetProfileForm } from './resetProfileForm';

export const makeResetProfileForm = action(() => {
  return (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    const { info } = getState().identity;
    if (!info) dispatch(updateProfileFormStatus('unavailable'));
    else dispatch(resetProfileForm(info));
  };
});
