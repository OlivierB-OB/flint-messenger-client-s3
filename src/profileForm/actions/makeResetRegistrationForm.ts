import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { resetProfileForm } from './resetProfileForm';

export function makeResetRegistrationForm() {
  return (dispatch: ThunkDispatch<IAppState, void, Action>) => {
    dispatch(resetProfileForm());
  };
}
