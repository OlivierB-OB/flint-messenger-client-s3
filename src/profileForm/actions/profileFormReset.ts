import { IProfileFormResetAction, PROFILE_FORM_RESET } from '../types';

export function profileFormReset(): IProfileFormResetAction {
  return { type: PROFILE_FORM_RESET };
}
