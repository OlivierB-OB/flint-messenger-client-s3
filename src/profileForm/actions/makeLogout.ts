import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { makeExitApplication } from '../../layout/actions/makeExitApplication';
import { config } from '../../config';

const { api_backend_url } = config;

export const makeLogout = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>) => {
    dispatch(updateProfileFormStatus('unavailable'));

    try {
      await axios.post(`${api_backend_url}/logout`, { withCredentials: true });
      dispatch(makeExitApplication());
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
