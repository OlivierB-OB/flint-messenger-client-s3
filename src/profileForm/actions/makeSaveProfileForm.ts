import axios from 'axios';
import { batch } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { updateIdentity } from '../../identity/actions/updateIdentity';
import { config } from '../../config';

const { api_backend_url } = config;

export const makeSaveProfileForm = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateProfileFormStatus('unavailable'));

    const { profileForm } = getState();
    const { lastName, firstName, password } = profileForm.fields;

    // FIXME validate form before sending

    try {
      const data = {
        lastName: lastName.value,
        firstName: firstName.value,
        password: password.value,
      };
      if (!password.value) delete data.password;
      const response = await axios.patch(`${api_backend_url}/profile`, data, { withCredentials: true });
      batch(() => {
        dispatch(updateIdentity(response.data));
        dispatch(updateProfileFormStatus('success'));
      });
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
