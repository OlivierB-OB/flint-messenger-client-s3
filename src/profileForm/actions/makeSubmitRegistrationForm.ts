import axios from 'axios';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { action } from '../../utils/action';
import { IAppState } from '../../appReducer';
import { history } from '../../history';
import { updateProfileFormStatus } from './updateProfileFormStatus';
import { makeInitializeApplication } from '../../layout/actions/makeInitializeApplication';
import { config } from '../../config';
import { makeExitApplication } from '../../layout/actions/makeExitApplication';

const { api_backend_url } = config;

export const makeSubmitRegistrationForm = action(() => {
  return async (dispatch: ThunkDispatch<IAppState, void, Action>, getState: () => IAppState) => {
    dispatch(updateProfileFormStatus('unavailable'));

    const { profileForm } = getState();
    const { email, lastName, firstName, password } = profileForm.fields;

    // FIXME validate form before sending

    try {
      const response = await axios.post(
        `${api_backend_url}/register`,
        {
          email: email.value,
          lastName: lastName.value,
          firstName: firstName.value,
          password: password.value,
        },
        { withCredentials: true },
      );
      await dispatch(makeExitApplication());
      dispatch(makeInitializeApplication(response.data));
      history.push(`/profile`);
    } catch (error) {
      dispatch(updateProfileFormStatus('error'));
    }
  };
});
