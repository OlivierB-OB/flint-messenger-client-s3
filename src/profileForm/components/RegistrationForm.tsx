import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import React from 'react';
import { connect } from 'react-redux';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { IAppState } from '../../appReducer';
import { updateProfileForm } from '../actions/updateProfileForm';
import { IProfileFormFields, IProfileFormStatus } from '../types';
import { IdentitySection } from './IdentitySection';
import { CredentialsSection } from './CredentialsSection';
import { Alert } from '../../layout/components/Alert';
import { makeResetRegistrationForm } from '../actions/makeResetRegistrationForm';
import { makeSubmitRegistrationForm } from '../actions/makeSubmitRegistrationForm';

export interface IRegistrationDisplayFormProps {
  status: IProfileFormStatus;
  fields: IProfileFormFields;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
  saveProfile(): void;
}

export function RegistrationFormDisplay({ status, fields, update, saveProfile }: IRegistrationDisplayFormProps) {
  const { email, firstName, lastName, password, confirmation } = fields;
  return (
    <Container>
      <Box style={{ margin: '2rem 0' }}>
        <Alert status={status} />
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="space-evenly" alignItems="flex-start">
          <Grid item xs={4}>
            <IdentitySection
              allowEmailEdition={true}
              email={email}
              firstName={firstName}
              lastName={lastName}
              update={update}
            />
          </Grid>
          <Grid item xs={4}>
            <CredentialsSection password={password} confirmation={confirmation} update={update} />
          </Grid>
        </Grid>
      </Box>
      <Box style={{ margin: '2rem 0' }}>
        <Grid container justify="flex-end">
          <Grid item xs={2}>
            <Button variant="contained" color="primary" fullWidth={true} onClick={saveProfile}>
              Register
            </Button>
          </Grid>
        </Grid>
      </Box>
    </Container>
  );
}

const mapStateToProps = ({ profileForm }: IAppState) => ({
  status: profileForm.status,
  fields: profileForm.fields,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  update: <T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']) =>
    dispatch(updateProfileForm(field, value)),
  resetForm: () => dispatch(makeResetRegistrationForm()),
  saveProfile: () => dispatch(makeSubmitRegistrationForm()),
});

export const RegistrationForm = connect(mapStateToProps, mapDispatchToProps)(RegistrationFormDisplay);
