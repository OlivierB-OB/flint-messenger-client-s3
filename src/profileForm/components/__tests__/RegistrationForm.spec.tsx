import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import { RegistrationForm } from '../RegistrationForm';
import { makeSubmitRegistrationForm } from '../../actions/makeSubmitRegistrationForm';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('RegistrationForm', () => {
  let Connected: any = RegistrationForm;
  let spySubmit: jest.Mock;
  let restoreSubmit: () => void;

  beforeEach(() => {
    [spySubmit, restoreSubmit] = makeSubmitRegistrationForm.mock(jest.fn());
  });

  afterEach(() => {
    restoreSubmit();
  });

  it('should provide status, fields and callbacks to RegistrationFormDisplay', async () => {
    const store = mockStore({
      profileForm: {
        status: 'profileForm status',
        fields: 'profileForm fields',
      },
    });
    const component = shallow(<Connected store={store} />).dive();
    expect(component).toMatchSnapshot();

    (component.prop('update') as any)('field to update', 'new value');
    expect(store.getActions()).toEqual([
      {
        field: 'field to update',
        value: 'new value',
        type: 'UPDATE_PROFILE_FORM',
      },
    ]);
    store.clearActions();

    (component.prop('saveProfile') as any)();
    expect(spySubmit).toHaveBeenCalled();
  });
});
