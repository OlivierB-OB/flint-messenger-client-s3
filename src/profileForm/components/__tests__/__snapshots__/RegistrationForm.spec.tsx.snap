// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm should provide status, fields and callbacks to RegistrationFormDisplay 1`] = `
<RegistrationFormDisplay
  fields="profileForm fields"
  saveProfile={[Function]}
  status="profileForm status"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  update={[Function]}
/>
`;
